datafusion/optimizer/src/rewrite_cycle.rs

```diff
@@ -0,0 +1,262 @@
+    /// logic of the loop. The function is expected to call [RewriteCycleState::rewrite] for each
+    /// individual [TreeNodeRewriter] in the cycle. [RewriteCycleState::rewrite] returns a [RewriteCycleControlFlow]
+    /// result, indicating whether the loop should break or continue.
+    ///
+    /// ```rust
+    ///
+    /// use arrow::datatypes::{Schema, Field, DataType};
+    /// use datafusion_expr::{col, lit};
+    /// use datafusion_common::{DataFusionError, ToDFSchema};
+    /// use datafusion_expr::execution_props::ExecutionProps;
+    /// use datafusion_expr::simplify::SimplifyContext;
+    /// use datafusion_optimizer::rewrite_cycle::RewriteCycle;
+    /// use datafusion_optimizer::simplify_expressions::{Simplifier, ConstEvaluator};
+    ///
+    /// // Create the schema
+    /// let schema = Schema::new(vec![
+    ///     Field::new("c1", DataType::Int64, true),
+    ///     Field::new("c2", DataType::UInt32, true),
+    /// ]).to_dfschema_ref().unwrap();
+    ///
+    /// // Create the rewriters
+    /// let props = ExecutionProps::new();
+    /// let context = SimplifyContext::new(&props)
+    ///    .with_schema(schema);
+    /// let mut simplifier = Simplifier::new(&context);
+    /// let mut const_evaluator = ConstEvaluator::try_new(&props).unwrap();
+    ///
+    /// // ((c4 < 1 or c3 < 2) and c3 < 3) and false
+    /// let expr = col("c2")
+    ///     .lt(lit(1))
+    ///     .or(col("c1").lt(lit(2)))
+    ///     .and(col("c1").lt(lit(3)))
+    ///     .and(lit(false));
+    ///
+    /// // run the rewrite cycle loop
+    /// let (expr, info) = RewriteCycle::new()
+    ///     .with_max_cycles(4)
+    ///     .each_cycle(expr, |cycle_state| {
+    ///         cycle_state
+    ///             .rewrite(&mut const_evaluator)?
+    ///             .rewrite(&mut simplifier)
+    ///     }).unwrap();
+    /// assert_eq!(expr, lit(false));
+    /// assert_eq!(info.completed_cycles(), 2);
+    /// assert_eq!(info.total_iterations(), 4);
+    /// ```
+    ///
+    pub fn each_cycle<
+        Node: TreeNode,
+        F: FnMut(
+            RewriteCycleState<Node>,
+        ) -> RewriteCycleControlFlow<RewriteCycleState<Node>>,
+    >(
+        &self,
+        node: Node,
+        mut f: F,
+    ) -> Result<(Node, RewriteCycleInfo)> {
+        let mut state = RewriteCycleState::new(node);
+        if self.max_cycles == 0 {
+            return state.finish();
+        }
+        // run first cycle then record number of rewriters
+        state = match f(state) {
+            ControlFlow::Break(result) => return result?.finish(),
+            ControlFlow::Continue(node) => node,
+        };
+        state.record_cycle_length();
+        if state.is_done() {
+            return state.finish();
+        }
+        // run remaining cycles
+        match (1..self.max_cycles).try_fold(state, |state, _| f(state)) {
+            ControlFlow::Break(result) => result?.finish(),
```

Comment:
should we return here?

Reply:
I think when we hit controlflow::break, it means `is_done`, so we can early return the result

```rust
ControlFlow::Break(result) => return result?.finish(),
```

Reply:
@erratic-pattern how about this?

Reply:
I tried this suggestion locally and it seems to work well 👍 

```suggestion
            ControlFlow::Break(result) => return result?.finish(),
```

Reply:
> I think when we hit controlflow::break, it means `is_done`, so we can early return the result
> 
> ```rust
> ControlFlow::Break(result) => return result?.finish(),
> ```

It is not needed because of implicit return semantics, since we are already at the end of the function. I can add an explicit return as it probably makes the code easier to read, but there may be some linter error that gets triggered

Reply:
you are right, `ControlFlow::Break` does the early return